File: harness/oracle/cases/recall-history.ts
/**
 * Oracle case #4: Commonplace Book recall anti-repetition ring buffer.
 *
 * Drives the REAL `parseRecallHistory`, `recentlyWhisperedIdSet`, and
 * `appendRecallTurn` from lib/memory/recall-history.ts over a fixed corpus of
 * arbitrary / deliberately-malformed JSON inputs and emits NDJSON. This is the
 * producer side of the `recentlyWhisperedIds` set that recall-tags.ts consumes.
 *
 * The module is pure + I/O-free (no DB, no logging) — its only job is to coerce
 * a `chats.commonplaceRecallHistory` JSON column (typed `unknown`) into a clean
 * `string[][]`, union it, and append+trim a ring buffer of the last
 * RECALL_HISTORY_TURNS whisper-turns.
 *
 * Because the inputs are arbitrary JSON (the whole point is the coercion of
 * malformed data), each row carries BOTH the input `raw` and the expected
 * output: the Rust differential test feeds the SAME `raw` (parsed to
 * serde_json::Value) through the port and compares. This avoids transcribing
 * subtle malformed inputs twice and guarantees both sides see identical bytes.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/recall-history.ts \
 *     > /tmp/oracle-recall-history.ndjson
 */

import {
  parseRecallHistory,
  recentlyWhisperedIdSet,
  appendRecallTurn,
  parseRetroSignatures,
  appendRetroSignature,
  type RecallHistory,
} from '@/lib/memory/recall-history';

// Row kinds — Rust dispatches on `kind`. Each carries the input `raw` verbatim.
type Row =
  | { kind: 'parse'; id: string; raw: unknown; out: string[][] }
  | { kind: 'set'; id: string; raw: unknown; out: string[] }
  | {
      kind: 'append';
      id: string;
      raw: unknown;
      newIds: string[];
      out: { turns: string[][]; retroSignatures?: string[] };
    }
  | { kind: 'parseSig'; id: string; raw: unknown; out: string[] }
  | {
      kind: 'appendSig';
      id: string;
      raw: unknown;
      newIds: string[];
      signature: string;
      out: { turns: string[][]; retroSignatures?: string[] };
    };

const rows: Row[] = [];

// ---------------------------------------------------------------------------
// parseRecallHistory — coercion of arbitrary/malformed JSON to string[][].
// ---------------------------------------------------------------------------
const parseCases: Array<[string, unknown]> = [
  ['null', null],
  ['number', 42],
  ['string', 'hi'],
  ['array-not-object', ['a']], // an array has no 'turns' key → []
  ['object-no-turns', { foo: 1 }],
  ['turns-null', { turns: null }],
  ['turns-string', { turns: 'x' }],
  ['turns-empty', { turns: [] }],
  ['simple', { turns: [['a', 'b'], ['c']] }],
  // drops non-string entries, empty strings, and a whole non-array "turn"
  ['drop-nonstring', { turns: [['a', 1, '', 'b'], 'notarray', ['c', null, true]] }],
  ['empty-inner-preserved', { turns: [[], ['a']] }], // empty inner array survives
  ['inner-all-dropped', { turns: [[1, 2, '']] }], // inner filters to [] but stays
  ['four-turns-no-trim', { turns: [['a'], ['b'], ['c'], ['d']] }], // parse does NOT trim
  ['nested-arrays-as-nonstring', { turns: [[['x']]] }], // inner element is an array → dropped
];
for (const [id, raw] of parseCases) {
  rows.push({ kind: 'parse', id, raw, out: parseRecallHistory(raw) });
}

// ---------------------------------------------------------------------------
// recentlyWhisperedIdSet — union (dedup) across retained turns. Emitted as the
// insertion-ordered [...set]; the Rust port returns a HashSet, so the test
// compares set membership (order-independent).
// ---------------------------------------------------------------------------
const setCases: Array<[string, unknown]> = [
  ['empty', { turns: [] }],
  ['dedup-across-turns', { turns: [['a', 'b'], ['b', 'c']] }],
  ['dedup-within', { turns: [['a', 'a', 'b']] }],
  ['with-malformed', { turns: [['a', 1, ''], ['b']] }],
  ['null', null],
];
for (const [id, raw] of setCases) {
  rows.push({ kind: 'set', id, raw, out: [...recentlyWhisperedIdSet(raw)] });
}

// ---------------------------------------------------------------------------
// appendRecallTurn — dedup newIds (no empty-string filter), push, trim last N.
// Empty newIds → no-op (returns parsed turns UNCHANGED, not trimmed).
// ---------------------------------------------------------------------------
const appendCases: Array<[string, unknown, string[]]> = [
  ['append-to-empty', null, ['a', 'b']],
  ['dedup-newids', { turns: [] }, ['a', 'a', 'b']],
  ['empty-newids-noop', { turns: [['a'], ['b']] }, []],
  ['empty-newids-no-trim', { turns: [['a'], ['b'], ['c'], ['d']] }, []], // over-cap, untrimmed
  ['trim-to-three', { turns: [['a'], ['b'], ['c']] }, ['d']], // oldest drops out
  ['newids-empty-string-kept', { turns: [] }, ['', 'a', 'a']], // append does NOT filter ''
  ['append-parses-malformed-raw', { turns: [['a', 1], 'x', ['b']] }, ['c']],
  // The retro-signature list rides the append untouched (both branches).
  ['carries-retro-signatures', { turns: [['a']], retroSignatures: ['s1', 's2'] }, ['b']],
  ['carries-retro-signatures-empty-newids', { turns: [['a']], retroSignatures: ['s1'] }, []],
  ['malformed-retro-signatures-dropped', { turns: [['a']], retroSignatures: ['', 3, 's1'] }, ['b']],
];
for (const [id, raw, newIds] of appendCases) {
  rows.push({ kind: 'append', id, raw, newIds, out: appendRecallTurn(raw, newIds) });
}

// ---------------------------------------------------------------------------
// parseRetroSignatures — the spam-guard list's coercion (same `unknown` column).
// ---------------------------------------------------------------------------
const parseSigCases: Array<[string, unknown]> = [
  ['null', null],
  ['number', 42],
  ['string', 'hi'],
  ['array-not-object', ['a']], // no 'retroSignatures' key → []
  ['object-no-key', { turns: [['a']] }],
  ['sigs-null', { retroSignatures: null }],
  ['sigs-string', { retroSignatures: 'x' }],
  ['sigs-empty', { retroSignatures: [] }],
  ['simple', { retroSignatures: ['s1', 's2'] }],
  ['drop-nonstring-and-empty', { retroSignatures: ['s1', 1, '', null, true, ['s2'], 's3'] }],
  ['four-no-trim', { retroSignatures: ['s1', 's2', 's3', 's4'] }], // parse does NOT trim
];
for (const [id, raw] of parseSigCases) {
  rows.push({ kind: 'parseSig', id, raw, out: parseRetroSignatures(raw) });
}

// ---------------------------------------------------------------------------
// appendRetroSignature — append (NO dedup) + trim to the last N. Runs on the
// already-appended history exactly as buildContext's persist does:
//   appendRetroSignature(appendRecallTurn(raw, newIds), signature)
// An empty signature is a no-op (the falsy guard) — notably NOT a trim.
// ---------------------------------------------------------------------------
const appendSigCases: Array<[string, unknown, string[], string]> = [
  ['first-signature', null, ['a'], 'w1|alice'],
  ['second-signature', { turns: [], retroSignatures: ['w1|alice'] }, ['a'], 'w2|bob'],
  ['duplicate-not-deduped', { retroSignatures: ['w1|alice'] }, [], 'w1|alice'],
  ['trim-to-three', { retroSignatures: ['s1', 's2', 's3'] }, [], 's4'],
  ['over-cap-input-trimmed-on-append', { retroSignatures: ['s1', 's2', 's3', 's4'] }, [], 's5'],
  ['empty-signature-noop', { retroSignatures: ['s1', 's2', 's3', 's4'] }, [], ''],
  ['empty-signature-noop-no-list', { turns: [['a']] }, ['b'], ''],
  ['malformed-list-coerced-first', { retroSignatures: [1, 's1', ''] }, ['a'], 's2'],
];
for (const [id, raw, newIds, signature] of appendSigCases) {
  const base: RecallHistory = appendRecallTurn(raw, newIds);
  rows.push({ kind: 'appendSig', id, raw, newIds, signature, out: appendRetroSignature(base, signature) });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/recall_history.rs
//! Port of lib/memory/recall-history.ts — the Commonplace Book recall
//! anti-repetition ring buffer.
//!
//! Persisted per-chat in `chats.commonplaceRecallHistory` as a JSON object
//! `{ turns: string[][] }` — one inner array of whispered memory IDs per recent
//! turn, most recent last, capped to [`RECALL_HISTORY_TURNS`]. (Wrapped in an
//! object rather than a bare array so it stays a plain JSON record column, not an
//! auto-detected SQLite "array column".)
//!
//! This is the producer side of the "recently whispered" set that
//! `recall_tags::recently_whispered_multiplier` consumes: the recall path unions
//! these IDs and applies a bounded penalty so the same memory doesn't get
//! whispered turn after turn and read as a stuck record.
//!
//! Pure + I/O-free. The TS parameter is typed `unknown` (an arbitrary, possibly
//! malformed JSON column), so the port takes `&serde_json::Value` and reproduces
//! the exact coercion: only a plain object with a `turns` array qualifies; each
//! turn must itself be an array; within a turn only non-empty strings survive.
//! Every malformed-input branch is preserved verbatim — that coercion is the
//! whole reason this module exists.

use std::collections::HashSet;

use serde_json::Value;

/// How many recent whisper-turns to remember for anti-repetition.
pub const RECALL_HISTORY_TURNS: usize = 3;

/// Persisted shape of the recall-history column.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RecallHistory {
    /// One inner array of whispered memory IDs per recent turn, most recent last.
    pub turns: Vec<Vec<String>>,
    /// Retro mini-recap signatures (episodic overhaul). Round 2 (P4.d13) only
    /// PRESERVES an existing list through [`append_recall_turn`] exactly as
    /// v4's `appendRecallTurn` does — the spam-guard machinery that writes and
    /// consumes them (`appendRetroSignature` / `RETRO_SIGNATURE_TURNS`) is the
    /// round-3 port.
    pub retro_signatures: Vec<String>,
}

impl RecallHistory {
    /// The persisted JSON shape (v4 `RecallHistory`): always `{ "turns": … }`,
    /// with `retroSignatures` present only when non-empty — key order matches
    /// v4's object literals.
    pub fn to_value(&self) -> Value {
        let mut obj = serde_json::Map::new();
        obj.insert(
            "turns".to_string(),
            Value::Array(
                self.turns
                    .iter()
                    .map(|t| Value::Array(t.iter().cloned().map(Value::String).collect()))
                    .collect(),
            ),
        );
        if !self.retro_signatures.is_empty() {
            obj.insert(
                "retroSignatures".to_string(),
                Value::Array(
                    self.retro_signatures
                        .iter()
                        .cloned()
                        .map(Value::String)
                        .collect(),
                ),
            );
        }
        Value::Object(obj)
    }
}

/// Coerce the persisted retro-signature list, dropping anything malformed (v4
/// `parseRetroSignatures`: only a plain object carrying `retroSignatures`
/// qualifies; only non-empty strings survive).
pub fn parse_retro_signatures(raw: &Value) -> Vec<String> {
    let sigs = match raw.as_object().and_then(|o| o.get("retroSignatures")) {
        Some(s) => s,
        None => return Vec::new(),
    };
    let sigs = match sigs.as_array() {
        Some(a) => a,
        None => return Vec::new(),
    };
    sigs.iter()
        .filter_map(|s| s.as_str())
        .filter(|s| !s.is_empty())
        .map(str::to_string)
        .collect()
}

/// Coerce the raw JSON column into a clean `Vec<Vec<String>>`, dropping anything
/// malformed.
///
/// Mirrors the TS `raw && typeof raw === 'object' && 'turns' in raw` guard: only
/// a plain JSON object carrying a `turns` key qualifies. A bare array does NOT
/// (`'turns' in [...]` is false in JS, and a `Value::Array` is not a
/// `Value::Object` here), null/number/string fall through, and a missing or
/// non-array `turns` yields an empty buffer. Within each turn, only strings of
/// length > 0 survive; a non-array turn element is dropped ENTIRELY (it does not
/// degrade to an empty turn), while an array turn that filters down to empty is
/// PRESERVED as `[]`.
pub fn parse_recall_history(raw: &Value) -> Vec<Vec<String>> {
    let turns = match raw.as_object().and_then(|o| o.get("turns")) {
        Some(t) => t,
        None => return Vec::new(),
    };
    let turns = match turns.as_array() {
        Some(a) => a,
        None => return Vec::new(),
    };
    turns
        .iter()
        .filter_map(|turn| turn.as_array()) // non-array turns dropped entirely
        .map(|turn| {
            turn.iter()
                .filter_map(|id| id.as_str())
                .filter(|s| !s.is_empty())
                .map(|s| s.to_string())
                .collect::<Vec<String>>()
        })
        .collect()
}

/// Union of all memory IDs across the retained recent turns.
///
/// The TS returns a `ReadonlySet<string>` (insertion-ordered, but consumers only
/// ever `.has()` it), so the natural Rust shape is a `HashSet<String>` — exactly
/// what `recall_tags::recently_whispered_multiplier` takes.
pub fn recently_whispered_id_set(raw: &Value) -> HashSet<String> {
    let mut set = HashSet::new();
    for turn in parse_recall_history(raw) {
        for id in turn {
            set.insert(id);
        }
    }
    set
}

/// Append this turn's whispered IDs and trim to the last [`RECALL_HISTORY_TURNS`].
///
/// Subtleties preserved from v4:
///   * Empty turns are not recorded — the buffer tracks recent *whispers*, not
///     silence, so a gap turn doesn't prematurely age a still-recent memory out
///     of the penalty. With empty `new_ids` the parsed buffer is returned
///     UNCHANGED — notably NOT trimmed, so an over-cap column survives a silent
///     turn intact (mirrors TS `return { turns }`).
///   * `new_ids` is only de-duplicated (first-occurrence order), NOT
///     empty-string-filtered — the `[...new Set(newIds)]` step trusts its caller
///     to pass real IDs, so an empty string in `new_ids` is pushed as-is.
pub fn append_recall_turn(raw: &Value, new_ids: &[String]) -> RecallHistory {
    let mut turns = parse_recall_history(raw);
    // v4 (8bf3cb5f): an existing retroSignatures list rides through the append
    // untouched (the write must not destroy a v4-written spam-guard list).
    let retro_signatures = parse_retro_signatures(raw);
    if new_ids.is_empty() {
        return RecallHistory {
            turns,
            retro_signatures,
        };
    }
    // Dedup new_ids preserving first-occurrence order (mirrors `new Set(newIds)`);
    // deliberately no empty-string filter here.
    let mut seen = HashSet::new();
    let mut deduped = Vec::new();
    for id in new_ids {
        if seen.insert(id.as_str()) {
            deduped.push(id.clone());
        }
    }
    turns.push(deduped);
    // slice(-RECALL_HISTORY_TURNS): keep the last N turns.
    if turns.len() > RECALL_HISTORY_TURNS {
        turns.drain(0..turns.len() - RECALL_HISTORY_TURNS);
    }
    RecallHistory {
        turns,
        retro_signatures,
    }
}
